pkg/controller/statefulset/stateful_set.go

```diff
@@ -235,70 +218,77 @@ func (psc *StatefulSetController) getPodsForStatefulSet(ps *apps.StatefulSet) ([
 	return result, nil
 }
 
-// getStatefulSetForPod returns the pet set managing the given pod.
-func (psc *StatefulSetController) getStatefulSetForPod(pod *v1.Pod) *apps.StatefulSet {
-	ps, err := psc.psStore.GetPodStatefulSets(pod)
+// getStatefulSetForPod returns the StatefulSet managing the given pod.
+func (ssc *StatefulSetController) getStatefulSetForPod(pod *v1.Pod) *apps.StatefulSet {
+	sets, err := ssc.setStore.GetPodStatefulSets(pod)
 	if err != nil {
 		glog.V(4).Infof("No StatefulSets found for pod %v, StatefulSet controller will avoid syncing", pod.Name)
 		return nil
 	}
-	// Resolve a overlapping statefulset tie by creation timestamp.
-	// Let's hope users don't create overlapping statefulsets.
-	if len(ps) > 1 {
-		glog.Errorf("user error! more than one StatefulSet is selecting pods with labels: %+v", pod.Labels)
-		sort.Sort(overlappingStatefulSets(ps))
+	// More than one set is selecting the same Pod
+	if len(sets) > 1 {
+		glog.Errorf("user error: more than one StatefulSet is selecting pods with labels: %+v", pod.Labels)
+		// The timestamp sort should not be necessary because we will enforce the CreatedBy requirement by
+		// name
+		sort.Sort(overlappingStatefulSets(sets))
+		// return the first created set for which pod is a member
+		for i := range sets {
+			if isMemberOf(&sets[i], pod) {
+				return &sets[i]
+			}
+		}
+		glog.V(4).Infof("No StatefulSets found for pod %v, StatefulSet controller will avoid syncing", pod.Name)
+		return nil
 	}
-	return &ps[0]
+	return &sets[0]
+
 }
 
 // enqueueStatefulSet enqueues the given statefulset in the work queue.
-func (psc *StatefulSetController) enqueueStatefulSet(obj interface{}) {
+func (ssc *StatefulSetController) enqueueStatefulSet(obj interface{}) {
 	key, err := controller.KeyFunc(obj)
 	if err != nil {
 		glog.Errorf("Cound't get key for object %+v: %v", obj, err)
 		return
 	}
-	psc.queue.Add(key)
+	ssc.queue.Add(key)
 }
 
 // worker runs a worker thread that just dequeues items, processes them, and marks them done.
 // It enforces that the syncHandler is never invoked concurrently with the same key.
-func (psc *StatefulSetController) worker() {
+func (ssc *StatefulSetController) worker() {
 	for {
 		func() {
```

Comment:
Mind extracting this into a separate function to follow what we are doing in other controllers? See the deployment controller for example - cc: @deads2k